src/CloudFlare/Api.php

```diff
@@ -47,24 +55,45 @@ class Api
     private $permissions = null;
 
     /**
+     * Default API endpoint
+     */
+    protected $endpoint = self::ENDPOINT_DEFAULT;
+
+    /**
      * Make a new instance of the API client
      * This can be done via providing the email address and api key as seperate parameters
      * or by passing in an already instantiated object from which the details will be extracted
+     *
+     * Signature 1:
+     * @param $arg1 Object client object
+     *
+     * Signature 2:
+     * @param $arg1 string hosting partner key
+     *
+     * Signature 3:
+     * @param $arg1 string user email
+     * @param $arg2 string user auth key
      */
-    public function __construct()
+    public function __construct($arg1, $arg2 =  null, $arg3 = null)
```

Comment:
Given that the provider authentication method is only used with the Provider class, I feel it would be simpler to revert this method back and just add another condition to the if statement for 3 arguments being passed in.

Then in the provider constructor you can do this:
```php
parent::__construct($key, null, 'host');
```
I know it's a bit of a workaround but it saves having variable such as $arg1, $arg2 etc.. and reduces the complexity of the conditions. It also means we aren't introducing the breaking change of the constructor requiring at least one argument.